nilearn/image/image.py

```diff
@@ -1892,3 +1897,514 @@ def copy_img(img):
+                        resample_to_first_img = True
+
+                if not _check_fov(niimg, ref_fov[0], ref_fov[1]):
+                    if target_fov is None:
+                        raise ValueError(
+                            f"Field of view of image #{i} is different from "
+                            "reference FOV.\n"
+                            f"Reference affine:\n{ref_fov[0]!r}\n"
+                            f"Image affine:\n{niimg.affine!r}\n"
+                            f"Reference shape:\n{ref_fov[1]!r}\n"
+                            f"Image shape:\n{niimg.shape!r}\n"
+                        )
+
+                    if resample_to_first_img:
+                        warnings.warn(
+                            "Affine is different across subjects."
+                            " Realignment on first subject "
+                            "affine forced",
+                            stacklevel=find_stack_level(),
+                        )
+                    niimg = cache(
+                        resample_img,
+                        memory,
+                        func_memory_level=2,
+                        memory_level=memory_level,
+                    )(
+                        niimg,
+                        target_affine=ref_fov[0],
+                        target_shape=ref_fov[1],
+                    )
+                yield niimg
+            except DimensionError as exc:
+                # Keep track of the additional dimension in the error
+                exc.increment_stack_counter()
+                raise
+            except TypeError as exc:
+                img_name = (
+                    f" ({niimg}) " if isinstance(niimg, (str, Path)) else ""
+                )
+
+                exc.args = (
+                    f"Error encountered while loading image #{i}{img_name}",
+                    *exc.args,
+                )
+                raise
+
+    # Raising an error if input generator is empty.
+    if i == -1:
+        raise ValueError("Input niimgs list is empty.")
+
+
+# ensure_ndim = 3 always returns a NiftiImage
+@overload
+def check_niimg(
+    niimg,
+    ensure_ndim: Literal[3] = ...,
+    atleast_4d: Literal[False] = ...,
+    dtype=...,
+    return_iterator: Literal[False] = ...,
+    wildcards=...,
+) -> Nifti1Image: ...
+
+
+# ensure_ndim = 4 with return_iterator=False always returns a NiftiImage
+@overload
+def check_niimg(
+    niimg,
+    ensure_ndim: Literal[4] = ...,
+    atleast_4d=...,
+    dtype=...,
+    return_iterator: Literal[False] = ...,
+    wildcards=...,
+) -> Nifti1Image: ...
+
+
+# ensure_ndim = 4 with return_iterator=True
+# always returns an iterator NiftiImage
+@overload
+def check_niimg(
+    niimg,
+    ensure_ndim: Literal[4] = ...,
+    atleast_4d=...,
+    dtype=...,
+    return_iterator: Literal[True] = ...,
+    wildcards=...,
+) -> Iterable[Nifti1Image]: ...
+
+
+@fill_doc
+def check_niimg(
```

Comment:
This function not only checks the image but also loads it. As an idea, would it be possible to rename it so that the name is more clear about its functionality (before making it public)?

Reply:
So `check_and load` ?

Reply:
I am thinking of packagers like myself who will need to patch older scientific code, which are unlikely to be ported, to be compatible with the new API. In order to limit the complexity of those patches, the best would be to limit changes to the import path only, and avoid function renaming.

That being said Hande's comment is fair, so the old names could be kept for legacy usage, whilst new ones with a better naming convention could be introduced, like `check_and_load` as proposed by Bertrand. Sounds to me like the best of both worlds.

Reply:
> with the old names) with a deprecation warning to give some time to adapt

which is very good!

But there won't be any adaptation from research code published today and no longer being actively maintained. So adaptation will come from downstream packagers eventually, and a patch like this is simpler to maintain:

```diff
-from nilearn._utils.niimg_conversion check_niimg
+from nilearn.image.image import check_niimg
```

than:

```diff
-from nilearn._utils.niimg_conversion check_niimg
+from nilearn.image.image import check_and_load

- check_niimg(...)
+ check_and_load(...)

- check_niimg(...)
+ check_and_load(...)

- check_niimg(...)
+ check_and_load(...)

...
```

My 2 cents.